File: 2025/day-02/part1.py
import sys

if len(sys.argv) > 1:
	INPUT = sys.argv[1]
else:
	INPUT = "input1.txt"

with open(INPUT) as f:
	ranges = f.read().split(",")
	ranges_nums = []
	for cur_range in ranges:
		rangesplit = cur_range.split("-")
		cur_range = [rangesplit[0], rangesplit[1]]
		ranges_nums.append(range(int(rangesplit[0]), int(rangesplit[1]) + 1))
	invalid_ids = 0
	for num_range in ranges_nums:
		for i in num_range:
			numstr = str(i)
			mid = int(len(numstr) / 2)
			left, right = numstr[0:mid], numstr[mid:]
			if left == right:
				print(numstr[0:mid], numstr[mid:])
				invalid_ids += i

print(invalid_ids)

File: 2025/day-02/src/part1.rs
#[tracing::instrument]
pub fn process(_input: &str) -> miette::Result<String> {
	let mut invalid_ids = 0;
	for range in _input.split(",") {
		let rangesplit: Vec<&str> = range.trim().split("-").collect();
		let (left, right) = (rangesplit[0], rangesplit[1]);
		let cur_range = [left, right].map(|cur| {
			cur.parse::<u64>()
				.expect("valid number with whitespace trimmed")
		});
		for i in cur_range[0]..=cur_range[1] {
			let numstr = i.to_string();
			let mid: usize = numstr.len() / 2;
			if numstr[..mid] == numstr[mid..] {
				invalid_ids += i;
			}
		}
	}
	Ok(invalid_ids.to_string())
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_process() -> miette::Result<()> {
		let input = "11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,446443-446449,38593856-38593862,565653-565659,824824821-824824827,2121212118-2121212124";
		assert_eq!("1227775554", process(input)?);
		Ok(())
	}
}
